packages/ra-core/src/controller/field/useReferenceFieldController.ts

```diff
@@ -33,31 +33,34 @@ export const useReferenceFieldController = <
     const createPath = useCreatePath();
     const resourceDefinition = useResourceDefinition({ resource: reference });
 
-    const result = useMemo(
-        () =>
-            ({
-                ...referenceRecordQuery,
-                link:
-                    referenceRecordQuery.referenceRecord != null
-                        ? link === false ||
-                          (link === 'edit' && !resourceDefinition.hasEdit) ||
-                          (link === 'show' && !resourceDefinition.hasShow)
-                            ? false
-                            : createPath({
-                                  resource: reference,
-                                  id: referenceRecordQuery.referenceRecord.id,
-                                  type:
-                                      typeof link === 'function'
-                                          ? link(
-                                                referenceRecordQuery.referenceRecord,
-                                                reference
-                                            )
-                                          : link,
-                              })
-                        : undefined,
-            }) as const,
-        [createPath, link, reference, referenceRecordQuery, resourceDefinition]
-    );
+    const result = useMemo(() => {
+        const defaultLink = resourceDefinition.hasEdit
+            ? 'edit'
+            : resourceDefinition.hasShow
+              ? 'show'
+              : false;
+        const isLinkFalse =
+            link === false || (link == null && defaultLink === false);
+        const linkString =
+            referenceRecordQuery.referenceRecord == null || isLinkFalse
+                ? false
+                : createPath({
+                      resource: reference,
+                      id: referenceRecordQuery.referenceRecord.id,
+                      // @ts-ignore TypeScript doesn't understand that type cannot be false here
+                      type:
+                          typeof link === 'function'
+                              ? link(
+                                    referenceRecordQuery.referenceRecord,
+                                    reference
+                                )
+                              : link ?? defaultLink,
+                  });
+        return {
+            ...referenceRecordQuery,
+            link: linkString,
+        } as const;
+    }, [createPath, link, reference, referenceRecordQuery, resourceDefinition]);
```

Comment:
We could probably extract this in a custom hook to use here and for the Datagrid rowClick